src/value.mts

```diff
@@ -720,77 +750,86 @@ export class ObjectValue extends Value {
+      // @ts-ignore
       m(this[s]);
     });
   }
 }
 
 export class ReferenceRecord {
+  Base: 'unresolvable' | Value;
+  ReferencedName: StringValue | SymbolValue | PrivateName;
+  Strict: BooleanValue;
+  ThisValue: ObjectValue | undefined;
   constructor({
     Base,
     ReferencedName,
     Strict,
     ThisValue,
-  }) {
+  }: Pick<ReferenceRecord, 'Base' | 'ReferencedName' | 'Strict' | 'ThisValue'>) {
     this.Base = Base;
     this.ReferencedName = ReferencedName;
     this.Strict = Strict;
     this.ThisValue = ThisValue;
   }
 
   // NON-SPEC
-  mark(m) {
+  mark(m: GCMarker) {
     m(this.Base);
     m(this.ReferencedName);
     m(this.ThisValue);
   }
 }
 
-export function Descriptor(O) {
-  if (new.target === undefined) {
-    return new Descriptor(O);
+// @ts-expect-error
+export function Descriptor(O: Pick<Descriptor, 'Configurable' | 'Enumerable' | 'Get' | 'Set' | 'Value' | 'Writable'>): Descriptor
+// @ts-expect-error
+export @callable() class Descriptor {
+  readonly Value?: Value;
+  readonly Get?: FunctionObjectValue;
+  readonly Set?: FunctionObjectValue;
+  readonly Writable?: BooleanValue;
+  readonly Enumerable?: BooleanValue;
+  readonly Configurable?: BooleanValue;
+  constructor(O: Pick<Descriptor, 'Configurable' | 'Enumerable' | 'Get' | 'Set' | 'Value' | 'Writable'>) {
+    this.Value = O.Value;
+    this.Get = O.Get;
+    this.Set = O.Set;
+    this.Writable = O.Writable;
+    this.Enumerable = O.Enumerable;
+    this.Configurable = O.Configurable;
+  }
+  everyFieldIsAbsent() {
+    return this.Value === undefined
+          && this.Get === undefined
+          && this.Set === undefined
+          && this.Writable === undefined
+          && this.Enumerable === undefined
+          && this.Configurable === undefined;
+  }
+  // NON-SPEC
+  mark(m: GCMarker) {
+    m(this.Value);
+    m(this.Get);
+    m(this.Set);
   }
-
-  this.Value = O.Value;
-  this.Get = O.Get;
-  this.Set = O.Set;
-  this.Writable = O.Writable;
-  this.Enumerable = O.Enumerable;
-  this.Configurable = O.Configurable;
 }
 
-Descriptor.prototype.everyFieldIsAbsent = function everyFieldIsAbsent() {
-  return this.Value === undefined
-    && this.Get === undefined
-    && this.Set === undefined
-    && this.Writable === undefined
-    && this.Enumerable === undefined
-    && this.Configurable === undefined;
-};
-
-// NON-SPEC
-Descriptor.prototype.mark = function mark(m) {
-  m(this.Value);
-  m(this.Get);
-  m(this.Set);
-};
-
 export class DataBlock extends Uint8Array {
-  constructor(sizeOrBuffer, ...restArgs) {
+  constructor(sizeOrBuffer: number | ArrayBuffer, byteOffset?: number, length?: number) {
     if (sizeOrBuffer instanceof ArrayBuffer) {
-      // fine.
-      super(sizeOrBuffer, ...restArgs);
+      super(sizeOrBuffer, byteOffset, length);
     } else {
       Assert(typeof sizeOrBuffer === 'number');
       super(sizeOrBuffer);
     }
   }
 }
 
-export function Type(val) {
+export function Type(val: Value) {
```

Comment:
this annotation is not correct?

Reply:
No idea. Does the `Type` in the spec accept any other type than an ECMAScript language value?

Reply:
Not since https://github.com/tc39/ecma262/pull/2874 was merged, I think.

Reply:
Since https://github.com/tc39/ecma262/issues/2870 prefers Type(x) only accept ES language value, I will keep this signature, and if there is any new type error, it might indicate we find a something we need to change in the spec.

Reply:
regardless of the spec what I meant is that it doesn't match the code in the body of the function. I'm surprised ts didn't pick up on that somehow?

Reply:
![image](https://user-images.githubusercontent.com/5390719/207374851-60ddf938-3448-4beb-9231-b05027ffdd86.png)
yeah, typescript allows this 🤣